repositories/strategies/routingStrategy.js

```diff
@@ -0,0 +1,287 @@
+  trx,
+  { conditionId, optionId, checked }
+) => {
+  const table = trx("Routing_ConditionValues");
+  const existing = await table.where(toDb({ conditionId, optionId }));
+  if (checked || isEmpty(existing)) {
+    return table
+      .insert(toDb({ conditionId, optionId }))
+      .returning("*")
+      .then(head)
+      .then(fromDb);
+  } else {
+    return table
+      .where(toDb({ conditionId, optionId }))
+      .del()
+      .returning("*")
+      .then(head)
+      .then(fromDb);
+  }
+};
+
+async function getAvailableRoutingDestinations(trx, pageId) {
+  const page = await getPageById(trx, pageId);
+  const section = await getSectionById(trx, page.sectionId);
+  const pageDestinations = await getPageDestinations(trx, page);
+  const sectionDestinations = await getSectionDestinations(trx, section);
+  return concat(pageDestinations, sectionDestinations);
+}
+
+function checkRoutingDestinations(
+  availableRoutingDestinations,
+  sectionDestination,
+  pageDestination
+) {
+  const partitionedDestinations = partition(
+    availableRoutingDestinations,
+    "sectionId"
+  );
+  const questionPages = first(partitionedDestinations);
+  const sections = last(partitionedDestinations);
+
+  if (isNil(pageDestination)) {
+    if (!find(sections, { id: parseInt(sectionDestination) })) {
+      throw new Error(
+        `Unable to route from this question to section ${sectionDestination}`
+      );
+    }
+  } else if (!find(questionPages, { id: parseInt(pageDestination) })) {
+    throw new Error(
+      `Unable to route from this question to page ${pageDestination}`
+    );
+  }
+}
+
+async function createRoutingConditionStrategy(trx, routingCondition) {
+  const { questionPageId, answerId } = routingCondition;
+  let firstAnswerOnPage;
+  if (isNil(answerId)) {
+    firstAnswerOnPage = await getFirstAnswer(trx, questionPageId);
+  }
+
+  return insertRoutingCondition(trx, {
+    ...routingCondition,
+    answerId: get(firstAnswerOnPage, "id", answerId)
+  });
+}
+
+const intOrNull = value => (isNil(value) ? null : parseInt(value));
+
+async function createRoutingRuleStrategy(
+  trx,
+  { routingRuleSetId, operation, goto }
+) {
+  const sectionDestination = get(goto, "sectionId", null);
+  const pageDestination = get(goto, "pageId", null);
+
+  const routingRule = await insertRoutingRule(trx, {
+    operation,
+    routingRuleSetId,
+    sectionDestination: intOrNull(sectionDestination),
+    pageDestination: intOrNull(pageDestination)
+  });
+
+  const routingRuleSet = await getRoutingRuleSetById(
+    trx,
+    routingRule.routingRuleSetId
+  );
+  await createRoutingConditionStrategy(trx, {
+    comparator: "Equal",
+    routingRuleId: routingRule.id,
+    questionPageId: routingRuleSet.questionPageId
+  });
+
+  return routingRule;
+}
+
+async function createRoutingRuleSetStrategy(trx, questionPageId) {
+  const existingRuleSet = await findRoutingRuleSet(trx, questionPageId);
+  if (!isNil(existingRuleSet)) {
```

Comment:
@WickyNilliams How do you mean?

Reply:
i mean can you use a unique constraint in the DB to ensure that you can't have more than one rule set? since this code checks for an existing one then throws, can we have the DB enforce it for us instead?

Reply:
Oh I see, yeah that makes sense. The only reason I can see for not doing that is that RoutingRuleSets have an `isDeleted` column meaning they are intended for soft deletion.

In that case it's plausible you might have more than one RoutingRuleSet with the same page Id. In practice, we have no way to delete the routing rule set at the moment. Depends if deleting rule sets (i.e. removing all routing rules) becomes a thing?